### PR TITLE
Compare previous latest deployment id to current deployment id

### PR DESCRIPTION
**Pull request by nturinski**
Grabs the old latest deployment id before the deployment begins.

As I implemented this, I questioned how useful this actually is.  While this guarantees that deployments were successful, it doesn't guarantee that it was built successfully so deployments can still fail silently since technically, the deployment DID succeed, but the build didn't.

If a user tries to deploy a Local Git repo that is already deployed, it will also result in an error since the ids will be the same.  Should we consider that by design?

### CODE REVIEW

File: appservice/src/SiteWrapper.ts [hunk 1]
```diff
@@ -183,14 +183,24 @@ export class SiteWrapper {
 
     public async deploy(fsPath: string, client: WebSiteManagementClient, outputChannel: vscode.OutputChannel, configurationSectionName: string, confirmDeployment: boolean = true): Promise<void> {
         const config: SiteConfigResource = await this.getSiteConfig(client);
+        const kuduClient: KuduClient = await this.getKuduClient(client);
+        const deploymentId: string = 'latest';
+        let previousDeploymentId: string;
+        try {
+            previousDeploymentId = (await kuduClient.deployment.getResult(deploymentId)).id;
+            // current active deployment used to verify new deployment was successful
+        } catch (error) {
```
Comment (StephenWeatherford) [inline appservice/src/SiteWrapper.ts]: Perhaps you should check that the error is what you're expecting for this scenario?

File: appservice/src/SiteWrapper.ts [hunk 2]
```diff
@@ -183,14 +183,24 @@ export class SiteWrapper {
 
     public async deploy(fsPath: string, client: WebSiteManagementClient, outputChannel: vscode.OutputChannel, configurationSectionName: string, confirmDeployment: boolean = true): Promise<void> {
         const config: SiteConfigResource = await this.getSiteConfig(client);
+        const kuduClient: KuduClient = await this.getKuduClient(client);
+        const deploymentId: string = 'latest';
+        let previousDeploymentId: string;
```
Comment (StephenWeatherford) [inline appservice/src/SiteWrapper.ts]: string|undefined

File: appservice/src/SiteWrapper.ts [hunk 3]
```diff
@@ -537,7 +547,9 @@ export class SiteWrapper {
             await new Promise((resolve: () => void): void => { setTimeout(resolve, pollingInterval); });
             deployment = await kuduClient.deployment.getResult(deploymentId);
         }
-
+        if (deployment.id === previousDeploymentId) {
+            throw new Error(localize('unsuccessfulDeployment', 'Deployment was unsuccessful and fell back to last active deployment.'));
```
Comment (StephenWeatherford) [inline appservice/src/SiteWrapper.ts]: How about something like:

  The deployment was unsuccessful and fell back to the last active deployment. If this is a local git deployment, it could also mean this repo has already been deployed.

Comment (EricJizbaMSFT) [reply appservice/src/SiteWrapper.ts]: The deployment didn't actually fall back to the last deployment, though. If you navigate to the web app's url, you will see the broken app that you just deployed (you will _not_ see the last successful deployment).

The only thing that "fell back" is the deployment metadata reported by kudu when you query for 'latest'. But that's just an internal implementation detail to kudu and the user doesn't care about it.

In other words, I don't think we should mention "falling back" at all

File: appservice/src/SiteWrapper.ts (hunk 1 above)
Comment (EricJizbaMSFT) [reply appservice/src/SiteWrapper.ts]: @StephenWeatherford I just talked to Nathan offline and I think we should swallow all errors here. No matter if or how it fails, we should still move on and try to deploy IMO

File: appservice/src/SiteWrapper.ts [hunk 4]
```diff
@@ -537,7 +547,9 @@ export class SiteWrapper {
             await new Promise((resolve: () => void): void => { setTimeout(resolve, pollingInterval); });
             deployment = await kuduClient.deployment.getResult(deploymentId);
         }
-
+        if (deployment.id === previousDeploymentId) {
```
Comment (EricJizbaMSFT) [inline appservice/src/SiteWrapper.ts]:
```
if (previousDeploymentId !== undefined && deployment.id === previousDeploymentId)
```

File: appservice/src/SiteWrapper.ts (hunk 1 above)
Comment (nturinski) [reply appservice/src/SiteWrapper.ts]: Discussing with Eric, it may not be best to check for the specific error message we're expecting.

If the message ever changes or is localized, this will break the deployment.
If we get an error at this point of the deployment, this shouldn't break the deployment.  This is only for validation and we account for undefined ids.

If there is an error, it's better for it to be caught in the specific deployment command rather than during the check for a latest id.

File: appservice/src/SiteWrapper.ts (hunk 4 above)
Comment (nturinski) [reply appservice/src/SiteWrapper.ts]: Is this necessary?  If the previousDeploymentId is undefined, this will not be true unless deployment.id is also undefined which would probably be a failure anyway.

Comment (EricJizbaMSFT) [reply appservice/src/SiteWrapper.ts]: True - I didn't think about it that way. I might still prefer it for the sake of readability, but no strong preference

Comment (nturinski) [reply appservice/src/SiteWrapper.ts]: 👑
